File: 机试/北师大/北师大夏令营机试测试/问题D-汉诺塔.py
n, q = list(map(int, input().split()))

def qpow(x, y, m):
    ret = 1
    while y:
        if y & 1:
           ret = ret * x % m
        x = x * x % m
        y >>= 1
    return ret

print((qpow(3, n, q)-1) % q)

File: 机试/北师大/北师大夏令营机试测试/问题D-汉诺塔.cpp
#include <bits/stdc++.h>

using namespace std;
typedef long long ll;

// int dfs(ll n)
ll qpow(ll a, ll b, ll q) {
    ll ret=1;
    while (b) {
        if (b & 1) {ret = ret * a % q;}
        a = a * a % q;
        b >>= 1;
    }
    return ret;
}

ll hanoi(ll n, ll q) {
    if (n == 1) {return 2;}    
    return (hanoi(n-1, q) * 3 + 2) % q;
}

int main() {
    ll n, q;
    cin >> n >> q;
    // if (n <= 0 || q <= 0) {cout << 0;}
    cout << (qpow(3, n, q) - 1) % q << endl;
    return 0;
}
